## Intro to this study
This research study dove into understanding the gut microbiome of covid positive patients at different severity levels using long-read 16S sequencing with the following parameters:

* ONT R9.4.1 chemistry
* Dorado v0.5.0 super accuracy basecalling 
* Emu v3.4.5 taxonomic classification and abundance estimation 

Along with the long-read 16S sequencing data, there is a **plethora** of metadata that goes alongside each deidentified patient. We intend and encourage folks to go through this data further and use this vignette as a tool for processing further 16S long-read datasets using R.

All the code used in the analysis and visualization of Emu microbiome outputs can be found in our [documentation](https://villapollab.github.io/covid_biome/).
